packages/cli/src/util/dev/server.ts

```diff
@@ -1566,9 +1567,31 @@ export default class DevServer {
 
             // Retain orginal pathname, but override query parameters from the rewrite
             const beforeRewriteUrl = req.url || '/';
-            const rewriteUrlParsed = url.parse(beforeRewriteUrl);
-            rewriteUrlParsed.search = url.parse(rewritePath).search;
-            req.url = url.format(rewriteUrlParsed);
+
+            if (isURL(rewritePath)) {
+              const rewriteUrlParsed = new URL(rewritePath);
+              const rewriteOrigin = `${rewriteUrlParsed.protocol}//${rewriteUrlParsed.host}`;
+
+              // `this.address` already has localhost normalized from ip4 and ip6 values
+              const devServerHost = new URL(this.address).host;
+              const isSameOrigin = devServerHost === rewriteUrlParsed.host;
+
+              if (isSameOrigin) {
+                req.url = rewritePath.replace(rewriteOrigin, '');
```

Comment:
Might be better to do a `rewritePath.slice(rewriteOrigin.length)` since you know the shape of the string already.

Reply:
Updated in [`0e9013a` (#8457)](https://github.com/vercel/vercel/pull/8457/commits/0e9013a125a62c331e088fb11c81b09d19e93b55)